cirq-core/cirq/circuits/circuit_operation_test.py

```diff
@@ -81,6 +81,17 @@ def test_repetitions_and_ids_length_mismatch():
         _ = cirq.CircuitOperation(circuit, repetitions=2, repetition_ids=['a', 'b', 'c'])
 
 
+def test_is_measurement_memoization():
+    a = cirq.LineQubit(0)
+    circuit = cirq.FrozenCircuit(cirq.measure(a, key='m'))
+    c_op = cirq.CircuitOperation(circuit)
+    assert cirq.is_measurement(c_op)
+    assert c_op._is_measurement_() is NotImplemented
+    # Memoize `has_measurements` in the circuit.
+    assert circuit.has_measurements()
+    assert c_op._is_measurement_() is True
```

Comment:
This is a red flag to me - memoization should not change the public properties of an object.

Reply:
True, but `_is_measurement_` isn't really a public property. `cirq.is_measurement` still gives the same answer (line 88) before memoization.
`_is_measurement_` is just the first (assumed to be efficient) strategy in calculating the protocol result. It should be allowed to return `NotImplemented` and let the protocol fallback to alternate strategies (i.e. decomposition).

Reply:
See my other comment - not memoizing in `_is_measurement_` prevents it from becoming efficient unless the user calls `has_measurement`, which is inconvenient behavior.

Reply:
Makes sense. See my other reply for the caveat in this approach.